### PR TITLE
Add flat mode for filemanager content provider, add last modified column

### PR DESCRIPTION
**Pull request by AObuchow**
allow changing sorting method by clicking on a column

Signed-off-by: Andrew Obuchowicz <aobuchow@redhat.com>

### CODE REVIEW

File: com.aobuchow.sample.commander/src/com/aobuchow/sample/commander/editor/FileManagerViewer.java [hunk 1]
```diff
@@ -101,15 +145,105 @@ public static String humanReadableByteCountSI(long bytes) {
...
+	private TableViewerColumn createColumnFor(TableViewer viewer, String label, int width) {
 		TableViewerColumn column = new TableViewerColumn(viewer, SWT.NONE);
-		column.getColumn().setWidth(400);
+		column.getColumn().setWidth(width);
 		column.getColumn().setText(label);
 		column.getColumn().setMoveable(true);
 		return column;
 	}
+	
+	// TODO: Refactor all of these by providing a method to run for the three comparison cases?
```
Comment (AObuchow) [inline com.aobuchow.sample.commander/src/com/aobuchow/sample/commander/editor/FileManagerViewer.java:156]: Refactor comparators

File: com.aobuchow.sample.commander/src/com/aobuchow/sample/commander/editor/FileManagerContentProvider.java [hunk 2]
```diff
@@ -13,28 +16,33 @@
...
-			return createModel((IResource) inputElement);
+			return createModel((IResource) inputElement, flatten);
 		} catch (CoreException e) {
 			e.printStackTrace();
 		}
 
 		return null;
 	}
 
-	// TODO: Refactor this
```
Comment (AObuchow) [inline com.aobuchow.sample.commander/src/com/aobuchow/sample/commander/editor/FileManagerContentProvider.java:31]: Still needs to be refactored. Refactor it or leave the comment

File: com.aobuchow.sample.commander/src/com/aobuchow/sample/commander/editor/FileManagerContentProvider.java [hunk 3]
```diff
@@ -13,28 +16,33 @@
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.jface.viewers.IStructuredContentProvider;
 
+import com.aobuchow.sample.commander.AudioPlayer;
 import com.aobuchow.sample.commander.resources.AudioFile;
 
 public class FileManagerContentProvider implements IStructuredContentProvider {
+	boolean flatten;
+	public FileManagerContentProvider(boolean flatten) {
```
Comment (AObuchow) [inline com.aobuchow.sample.commander/src/com/aobuchow/sample/commander/editor/FileManagerContentProvider.java:24]: Still needs a  toolbar button to toggle

Comment (AObuchow) [reply com.aobuchow.sample.commander/src/com/aobuchow/sample/commander/editor/FileManagerContentProvider.java:24]: Drag support & pasting should be disabled when the toggle is active

**Comment by AObuchow** [comment]
Still needs some work.

When a directory is flattened and it has lots of children, it takes time to generate the model.
In this case, a UI job or something should be run to create the model in async fashion.

@ingomohr have you ever encountered such an issue? (Viewer input/model creation blocking the UI)

**Comment by AObuchow** [comment]
Clicking on a column that is currently being used to sort should make the sorting happen in reverse.
The sorting column's name should have a ▲ or ▼ to indicate the sorting direction.

**Comment by ingomohr** [comment]
> Still needs some work.
> 
> When a directory is flattened and it has lots of children, it takes time to generate the model.
> In this case, a UI job or something should be run to create the model in async fashion.
> 
> @ingomohr have you ever encountered such an issue? (Viewer input/model creation blocking the UI)



> Still needs some work.
> 
> When a directory is flattened and it has lots of children, it takes time to generate the model.
> In this case, a UI job or something should be run to create the model in async fashion.
> 
> @ingomohr have you ever encountered such an issue? (Viewer input/model creation blocking the UI)

Yes. In those cases we (at work) usually use the Jobs API to run a process outside of the UI thread (so the UI isn't broken/frozen). For longer-running jobs you could also show some progress info.

**Comment by AObuchow** [comment]
@ingomohr okay awesome thanks, I'll have to run a job with a progress monitor when the user enables "flat directory mode" :)

**Comment by ingomohr** [comment]
> @ingomohr okay awesome thanks, I'll have to run a job with a progress monitor when the user enables "flat directory mode" :)

Yes, 👍😊 I would do that.

**Comment by ingomohr** [comment]
You could also/in addition track changes on the model (e.g. via Observer pattern, EMF adapter (if you want to use EMF) or something in that direction - and trigger a refresh of the UI when the model changed (can make sense for long running jobs that change a model displayed in the UI).

**Comment by AObuchow** [comment]
I might need a deferred content provider as I think the blocking is happening in the model creation.

This might help: https://www.codeaffine.com/2014/12/02/deferred-fetching-with-jface-viewers/

**Comment by AObuchow** [comment]
@ingomohr I think I'm going to merge this for now and make the async content provider/ui job another bug :)

**Pull request merged**